---
title: "MAGERI benchmark using reference standard DNA library"
author: "Mikhail Shugay"
date: "December 1, 2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load metadata

```{r}
library(stringr)
library(pROC)
library(ggplot2)
library(ggbeeswarm)
library(dplyr)
df.vmeta <- read.table("hd734_variant_metadata.txt", sep="\t", header=T) # variants observed in HD734 and covered by our primers
df.smeta <- read.table("sample_metadata.txt", sep="\t", header=T) %>% filter(type == "blank") # metadata for amplicon sequencing samples
```

VCF parsing function

```{r}
read_vcf <- function(file_name) {
  .vcf <- read.table(file_name, header = F, sep = "\t", stringsAsFactors = F)
  colnames(.vcf) <- c("chromosome", "position",	"skip1", "from", "to", "qual", "skip2", "info", "skip3", "skip4")
  .vcf$skip1 <- NULL
  .vcf$skip2 <- NULL
  .vcf$skip3 <- NULL
  .vcf$skip4 <- NULL
  
  .vcf$qual <- as.integer(.vcf$qual)
  .vcf <- subset(.vcf, nchar(from) == 1 &
                   nchar(to) == 1 & !is.na(qual)) # no indels
  
  .infosplit <- str_split_fixed(.vcf$info, regex("[=;]"), 9)[,c(2, 4, 8)]
  
  .vcf$coverage <- as.numeric(.infosplit[,1])
  .vcf$frequency <- as.numeric(.infosplit[,2])
  .vcf$cqs <- as.numeric(.infosplit[,3])
  .vcf$info <- NULL
  
  .vcf$count <- as.integer(round(.vcf$coverage * .vcf$frequency))
  
  .vcf
}

head(read_vcf("p126.h4_2_ballast_m1.vcf"))
```

Read samples with HD734 standard DNA and control human DNA, append metadata

```{r}
df <- data.frame()

read_vcf_with_metadata <- function(file_name, primer_set, replica, ratio, type) {
  .vcf <- read_vcf(file_name)
  .vcf <- merge(.vcf, df.vmeta, all.x = type != "standard", all.y = F)
  .vcf$known.frequency <- .vcf$known.frequency * ratio
  .vcf$known.frequency[is.na(.vcf$known.frequency)] <- 0
  .vcf$primer_set <- primer_set
  .vcf$replica <- primer_set
  .vcf$type <- type
  
  .vcf <- subset(.vcf, frequency < 0.4 & count > 0) # remove alleles in control
  
  .vcf
}

for (i in 1:nrow(df.smeta)) {
  df <- with(df.smeta, rbind(df, 
                             read_vcf_with_metadata(paste(prefix[i], "vcf", sep="."), 
                                                    primer_set[i],
                                                    replica[i],
                                                    ratio[i],
                                                    type[i])))
}
```

Histogram of CQS scores

```{r}
fig13 <- ggplot(df, aes(cqs)) +
  geom_histogram(fill="red", binwidth=1) +
  ylab("") + scale_x_continuous("Consensus quality score", limits=c(0, 41)) +
  theme_bw()
fig13
```


```{r}
p_err <- 1-pbinom(3, size = 5, p=1e-2) + 0.5*dbinom(3, size = 5, p=1e-2)
n_migs <- 100000
df.1 <- data.frame(count = df$frequency * n_migs, type = "Observed")
lambda <- p_err * n_migs
df.1 <- rbind(df.1, data.frame(count = rpois(nrow(df), lambda), type="Expected"))

fig14 <- ggplot(df.1, aes(x=count, fill=type)) +
  geom_histogram() +
  ylab("") + scale_x_log10("Error count per 100,000 MIGs") +
  scale_fill_brewer("Error type", palette = "Set1") +
  theme_bw()
fig14
```

